src/App.tsx

```diff
@@ -1,223 +1,225 @@
+import {
+  useContext,
+  useEffect,
+  useRef,
+  useState,
+  FC,
+  useMemo,
+  useCallback,
+} from 'react';
+import {
+  getTodos,
+  createTodo,
+  deleteTodo,
+  updateTodo,
+} from './api/todos';
 import { AuthContext } from './components/Auth/AuthContext';
+import { ErrorNotification } from './components/ErrorNotification';
+import { Footer } from './components/Footer';
+import { Header } from './components/Header';
+import { TodoList } from './components/TodoList';
+import { Todo } from './types/Todo';
+import { FilterType } from './types/FilterType';
+
+export const App: FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [errorMessage, setErrorMessage] = useState<string>('');
+  const [
+    completedFilter,
+    setCompletedFilter,
+  ] = useState<FilterType>(FilterType.ALL);
+  const [title, setTitle] = useState<string>('');
+  const [isAdding, setIsAdding] = useState<boolean>(false);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [deletingTodoIds, setDeletingTodoIds] = useState<number[]>([]);
+  const [updatingTodosIds, setUpdatingTodosIds] = useState<number[]>([]);
 
-export const App: React.FC = () => {
-  // eslint-disable-next-line @typescript-eslint/no-unused-vars
   const user = useContext(AuthContext);
   const newTodoField = useRef<HTMLInputElement>(null);
 
   useEffect(() => {
+    if (user) {
+      getTodos(user.id)
+        .then(setTodos)
+        .catch(() => {
+          setErrorMessage('Unable to load todos');
+        });
+    }
+
     // focus the element with `ref={newTodoField}`
     if (newTodoField.current) {
       newTodoField.current.focus();
     }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
+  const updatingTodo = useCallback(async (
+    todoId: number,
+    fieldsToUpdate: Partial<Pick<Todo, 'title' | 'completed'>>,
+  ) => {
+    setUpdatingTodosIds(prevIds => {
+      if (!prevIds.includes(todoId)) {
+        return [...prevIds, todoId];
+      }
+
+      return prevIds;
+    });
+
+    try {
+      await updateTodo(todoId, fieldsToUpdate);
+      setTodos(prevTodos => prevTodos.map(todo => {
+        if (todo.id !== todoId) {
+          return todo;
+        }
+
+        return Object.assign(todo, fieldsToUpdate);
+      }));
+    } catch {
+      setErrorMessage('Unable to update a todo');
+    } finally {
+      setUpdatingTodosIds(prevTodosIds => prevTodosIds.filter(id => (
+        id !== todoId)));
+    }
+  }, []);
+
+  const onDeleteTodo = useCallback((todoId: number) => {
+    setDeletingTodoIds(prev => [...prev, todoId]);
+    deleteTodo(todoId)
+      .then(() => (
+        setTodos(allTodos => allTodos.filter(todo => todo.id !== todoId))
+      ))
+      .catch(() => {
+        setErrorMessage('Unable to delete a todo');
+      })
+      .finally(() => (
+        setDeletingTodoIds(prev => prev.filter(id => id !== todoId))));
   }, []);
 
+  const onFormSubmit = useCallback(
```

Comment:
This callback actually adds todo
```suggestion
  const addTodo = useCallback(
```